proposals/20240606-jwt-handling.md

```diff
@@ -0,0 +1,135 @@
+    1. Backends that are not exposed to user namespaces (i.e. jupyter-web-app) are only reachable via the Istio IngressGateway.
+    2. The KFP backend [explicitly drops requests](https://github.com/kubeflow/manifests/blob/96ce068e16b2a707464471bddc0d2a58e403d1fc/apps/pipeline/upstream/base/installs/multi-user/istio-authorization-config.yaml#L37) from user namespaces if they have this header
+    3. In-cluster Pods that want to talk to `kubeflow` workloads, which understands identity, are [using a K8s ServiceAccount Token](https://www.kubeflow.org/docs/components/pipelines/v1/sdk/connect-api/#full-kubeflow-subfrom-inside-clustersub)
+### Limitations
+- Not able to express `AuthorizationPolicies` for group header in Istio
+- Limited possibility to use custom JWT claims as a source of information about authenticated user
+
+To accommodate the above limitations and improve the authnz flow in terms of security, maintenance and flexibility
+we propose to add the JWT to the `Authorization` header, so it can be digested by Istio and have the user details securely
+injected to auth headers when authorized. This will also enable us to define policies in the future for better handling of groups.
+https://istio.io/latest/docs/tasks/security/authorization/authz-jwt/
+
+But the above creates the following topics that require agreement on how to handle:
+1. There will be `id_tokens` from different issuers (i.e. from Dex, K8s) that the platform will need to handle
+2. Information of user is both in `kubeflow-userid` and in `id_token` of http request, for Kubeflow components to deduce the identity from
+3. It's not clear if backends should be validating the JWT (i.e. KFP right now validates ServiceAccount tokens [`1`](https://github.com/kubeflow/pipelines/blob/2.2.0/backend/src/apiserver/auth/authenticator_token_review.go#L47-L58) [`2`](https://github.com/kubeflow/pipelines/blob/2.2.0/backend/src/apiserver/resource/resource_manager.go#L1698-L1699) )
+
+## Spec
+This proposal aims to provide a uniform way for all backends to handle identity tokens, and agreeing on which levels of the stack should responsible for different parts.
+
+This spec proposes to standardise on the following high level agreement, for new backends:
+1. Requests hitting `kubeflow` apps, which expect user identity in requests, should have a JWT in `Authorization: Bearer <token>` header
+2. The service-mesh is responsible for validating the JWTs
+    1. The service-mesh must drop (401) a request if the JWT is invalid (`RequestAuthentication`)
+    2. The service-mesh must drop (403) a request if the JWT is not present, and the application expects requests to have a user identity (`AuthorizationPolicy`)
+3. The backends are not responsible for validating the JWTs or their existence
+4. The service-mesh must expose user and groups to `kubeflow-userid` and `kubeflow-groups` headers, after validating JWTs
```

Comment:
will it drop the bearer token header with the JWT before forwarding to the backend?

Reply:
No no the JWT will stay in the `Authorization` header, and the mesh will add the headers as an extra